### PR TITLE
Add UserFunctionCache to cache UDF's library

### PR DESCRIPTION
**Pull request by imay**
This patch replace LibCache with UserFunctionCache. LibCache use HDFS
URL to identify a UDF's Library, and when BE process restart all of
downloaded library should be loaded another time. We use function id
corresponding to a library, and when process restart, all downloaded
libraries can be loaded without another downloading.

### CODE REVIEW

File: be/src/runtime/user_function_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    entry->is_downloaded = true;
+
+    entry->ref();
+    _entry_map[function_id] = entry;
+
+    return Status::OK;
+}
+
+Status UserFunctionCache::_load_cached_lib() {
+    // craete library directory if not exist
```
[the remark this answers is not in the document]
Comment (imay) [reply be/src/runtime/user_function_cache.cpp]: my fault

File: be/src/runtime/user_function_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+            entry = it->second;
+        } else {
+            entry = new UserFunctionCacheEntry(
+                fid, checksum, _make_lib_file(fid, checksum));
+
+            entry->ref();
+            _entry_map.emplace(fid, entry);
+        }
+        entry->ref();
+    }
```
[the remark this answers is not in the document]
Comment (imay) [reply be/src/runtime/user_function_cache.cpp]: Yes, it should be called twice, one for entry map, one for current function.
release_entry should be added exprs.h

File: be/src/util/file_utils.cpp [hunk 3]
```diff
@@ -127,6 +127,43 @@ Status FileUtils::scan_dir(
...
+    if (dir == nullptr) {
+        char buf[64];
+        LOG(WARNING) << "fail to open dir, dir=" << dir_path << ", errmsg=" << strerror_r(errno, buf, 64);
+        return Status("fail to opendir");
+    }
+
+    struct dirent entry;
+    struct dirent* result = nullptr;
+    while (true) {
+        int ret = readdir_r(dir.get(), &entry, &result);
```
[the remark this answers is not in the document]
Comment (imay) [reply be/src/util/file_utils.cpp]: OK, I will change it

**Review by chenhao7253886 — APPROVED** [review]
LTMG

**Pull request merged**